File: java17/src/main/java/org/chelck/dp/Philosopher.java
package org.chelck.dp;

import java.util.Objects;
import java.util.Random;

public class Philosopher implements Runnable {
    private final BlockingTable table;
    private final int philosopherId;
    private int n;
    private final Random random;
    private final Timings timings;

    public Philosopher(BlockingTable table, int philosopherId, int n) {
        this.table = Objects.requireNonNull(table);
        this.philosopherId = philosopherId;
        this.n = n;
        this.random = new Random();
        this.timings = new Timings();
    }

    public void run() {
        System.out.println("Philosopher " + philosopherId + " is starting: ");
        try {
            timings.start();
            while (n-->0) {
                think();

                table.getLeftFork(philosopherId);
                table.getRightFork(philosopherId);

                eat();

                table.returnLeftFork(philosopherId);
                table.returnRightFork(philosopherId);
                timings.mark();
            }
        }
        catch (Exception e) {}
        System.out.println("Philosopher " + philosopherId + " is done: " + timings);
    }

    private void think() {
        long delay = random.nextLong(10);
        //System.out.println("Philosopher " + philosopherId + " is thinking for " + totalThinking);

//        try {
//            Thread.sleep(delay);
//        }
//        catch (InterruptedException e) {
//        }
    }

    private void eat() {
        long delay = random.nextLong(4);
        //System.out.println("Philosopher " + philosopherId + " is eating for " + totalEating);
        try {
            Thread.sleep(delay);
        }
        catch (InterruptedException e) {
        }
    }
}

File: scala3_mutex/src/main/scala/org/chelck/dp/Philosopher.scala
package org.chelck.dp

import java.util.{Objects, Random}

class Philosopher(val table: BlockingTable, val philosopherId: Int, var numberOfInterations: Int, var random: Random, var timings: Timings)
  extends Runnable {


  override def run(): Unit = {
    try {
      timings.start()
      (0 until numberOfInterations).foreach(_ -> {
        think()
        table.getLeftFork(philosopherId)
        table.getRightFork(philosopherId)
        eat()
        table.returnLeftFork(philosopherId)
        table.returnRightFork(philosopherId)
        timings.mark()
      })
    } catch {
      case e: Exception =>

    }
    println("Philosopher " + philosopherId + " is done: " + timings)
  }

  private def think(): Unit = {
  }

  private def eat(): Unit = {
    val delay = random.nextLong(2)
//    try Thread.sleep(delay)
//    catch {
//      case e: InterruptedException =>
//    }
  }
}
